numba/targets/boxing.py

```diff
@@ -541,10 +549,78 @@ def box_list(typ, val, c):
     return c.builder.load(res)
 
 
+class _NumbaTypeHelper(object):
+    """A helper for acquiring `numba.typeof` for type checking.
+
+    Usage
+    -----
+
+        # `c` is the boxing context.
+        with _NumbaTypeHelper(c) as nth:
+            # This contextmanager maintains the lifetime of the `numba.typeof`
+            # function.
+            the_numba_type = nth.typeof(some_object)
+            # Do work on the type object
+            do_checks(the_numba_type)
+            # Cleanup
+            c.pyapi.decref(the_numba_type)
+        # At this point *nth* should not be used.
+    """
+    def __init__(self, c):
+        self.c = c
+
+    def __enter__(self):
+        c = self.c
+        numba_name = c.context.insert_const_string(c.builder.module, 'numba')
+        numba_mod = c.pyapi.import_module_noblock(numba_name)
+        typeof_fn = c.pyapi.object_getattr_string(numba_mod, 'typeof')
+        self.typeof_fn = typeof_fn
+        c.pyapi.decref(numba_mod)
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        c = self.c
+        c.pyapi.decref(self.typeof_fn)
+
+    def typeof(self, obj):
+        res = self.c.pyapi.call_function_objargs(self.typeof_fn, [obj])
+        return res
+
+
 def _python_list_to_native(typ, obj, c, size, listptr, errorptr):
     """
     Construct a new native list from a Python list.
     """
+    def check_element_type(nth, itemobj):
+        typobj = nth.typeof(itemobj)
+        # Check if *typobj* is NULL
+        with c.builder.if_then(
+                cgutils.is_null(c.builder, typobj),
+                likely=False,
+                ):
+            c.builder.store(cgutils.true_bit, errorptr)
+            loop.do_break()
+        # Mandate that objects all have the same exact type
+        type_mismatch = c.builder.icmp_signed('!=', typobj, expected_typobj)
```

Comment:
the late bind of `expected_typobj` is unexpected!

Reply:
right, let's pass it in